### ISSUE TITLE
Relative path of root element is resolved against data-dir, not current directory

### ISSUE THREAD

**Issue by ProditorMagnus** [wesnoth/wesnoth#9092]
### Game and System Information

- **Version:** 1.19.0

### Description of the bug

When wesnoth -p is used on file which uses {./inclusion}, then preprocessing is done as if {inclusion} was written instead.

When {./inclusion} is in deeper level, then it works as it should.

### Steps to reproduce the behavior

When wesnoth -p is used on file which uses {./inclusion}, then preprocessing is done as if {inclusion} was written instead.
```
Ravana@RavanaE480 /cygdrive/c/Users/Ravana/Documents/My Games/Wesnoth1.18/data/add-ons/EventLoader/staging $ cat root.cfg
{./included.cfg}%

Ravana@RavanaE480 /cygdrive/c/Users/Ravana/Documents/My Games/Wesnoth1.18/data/add-ons/EventLoader/staging $ cat included.cfg
x=1%

Ravana@RavanaE480 /cygdrive/c/Users/Ravana/Documents/My Games/Wesnoth1.18/data/add-ons/EventLoader/staging $ wesnoth --log-error filesystem,preprocessor,cache,config,general --data-dir "C:\Users\Ravana\games\battle-for-wesnoth-1.18.0" --userdata-dir "C:\Users\Ravana\Documents\My Games\Wesnoth1.18" --preprocess-defines MULTIPLAYER,SKIP_CORE,__WML_TREE_TOOLS__ -p root.cfg out
Battle for Wesnoth v1.19.0-dev <unknown>
Started on Fri Jul 12 16:42:27 2024

Automatically found a possible data directory at: C:\Users\USER\games\battle-for-wesnoth-1.19.0
Starting with directory: 'C:\Users\USER\games\battle-for-wesnoth-1.18.0'
Now have with directory: 'C:\Users\USER\games\battle-for-wesnoth-1.18.0'
Overriding data directory with 'C:\Users\USER\games\battle-for-wesnoth-1.18.0'
'SKIP_CORE' defined.
added 4 defines.
skipped 'data/core'
preprocessing specified resource: root.cfg ...
20240712 16:42:27 error preprocessor: Macro/file './included.cfg' is missing
at root.cfg:1
Game error: Macro/file './included.cfg' is missing
at root.cfg:1
```

When {./inclusion} is in deeper level, then it works as it should.
```
Ravana@RavanaE480 /cygdrive/c/Users/Ravana/Documents/My Games/Wesnoth1.18/data/add-ons/EventLoader/staging $ cat subfolder.cfg
{~add-ons/EventLoader/staging/sub/root.cfg}%

Ravana@RavanaE480 /cygdrive/c/Users/Ravana/Documents/My Games/Wesnoth1.18/data/add-ons/EventLoader/staging $ cat sub/root.cfg
{./included.cfg}%

Ravana@RavanaE480 /cygdrive/c/Users/Ravana/Documents/My Games/Wesnoth1.18/data/add-ons/EventLoader/staging $ cat sub/included.cfg
x=2%

Ravana@RavanaE480 /cygdrive/c/Users/Ravana/Documents/My Games/Wesnoth1.18/data/add-ons/EventLoader/staging $ wesnoth --log-error filesystem,preprocessor,cache,config,general --data-dir "C:\Users\Ravana\games\battle-for-wesnoth-1.18.0" --userdata-dir "C:\Users\Ravana\Documents\My Games\Wesnoth1.18" --preprocess-defines MULTIPLAYER,SKIP_CORE,__WML_TREE_TOOLS__ -p subfolder.cfg out
Battle for Wesnoth v1.19.0-dev <unknown>
Started on Fri Jul 12 16:45:34 2024

Automatically found a possible data directory at: C:\Users\USER\games\battle-for-wesnoth-1.19.0
Starting with directory: 'C:\Users\USER\games\battle-for-wesnoth-1.18.0'
Now have with directory: 'C:\Users\USER\games\battle-for-wesnoth-1.18.0'
Overriding data directory with 'C:\Users\USER\games\battle-for-wesnoth-1.18.0'
'SKIP_CORE' defined.
added 4 defines.
skipped 'data/core'
preprocessing specified resource: subfolder.cfg ...
acquired 4 total defines.
preprocessing finished. Took 16 ticks.

Ravana@RavanaE480 /cygdrive/c/Users/Ravana/Documents/My Games/Wesnoth1.18/data/add-ons/EventLoader/staging $ cat out/subfolder.cfg
x=2
```

**Comment by soliton-** [comment]
The issue is what's considered the current dir.

With `wesnoth -p file.cfg out` `{CURRENT_DIR}` expands to nothing. And with `wesnoth -p path/file.cfg out` `{CURRENT_DIR}` expands to `path`.

CURRENT_DIR empty should mean the same as `.` i.e. the current dir the file is in but it currently appears to mean the game's `data/` dir which makes very little sense.

Apparently this was done as a confusing bug fix: https://github.com/wesnoth/wesnoth/commit/1c927c8407ec797ef9e9e58599a7b71e617fe864

If `current_dir` being empty is an issue then I don't understand why it was not made an error instead of returning a path that was clearly not asked for.

Anyway, even without touching the weird `get_wml_location` logic the preprocessor can just call `get_wml_location` with the dir argument set to "." instead.

**Comment by gfgtdf** [comment]
> Anyway, even without touching the weird get_wml_location logic the preprocessor can just call get_wml_location with the dir argument set to "." instead.

This wont work, if youpass `.` wesnoth will literally look for the file at "./included.cfg" (usually the working directory wesnoth was started from, or the folder the wesnoth.exe is i guess) which is most likely not the folder where where file.cfg is at. 

> Apparently this was done as a confusing bug fix: https://github.com/wesnoth/wesnoth/commit/1c927c8407ec797ef9e9e58599a7b71e617fe864

The "confusing bugfix" was a fix for a serious security problem that allowed umc code to read aribitiary files on the users computer.

To make it clear: `get_wml_location ` is the central place that translates "wesnoth" paths to actual paths on the hard drive and must always return a path that's safe to read from.

**Comment by gfgtdf** [comment]
> CURRENT_DIR empty should mean the same as `.` ..

im not fully sure here but iirc it current doesn't: In macros `.` is relative to the place that defines the macro but `CURRENT_DIR` is the directory that uses the macro.

**Comment by ProditorMagnus** [comment]
In case of command line use, the folder of parameter should be legal data root, no matter if it would be allowed during game or not.

**Comment by gfgtdf** [comment]
yeah. The root of the problem here is that the proproceccor code operates on "real" paths internally, fixing that would also also allow us to use {./../utils} in wml code.

**Comment by gfgtdf** [comment]
hmm okay i just saw that in your example in the first code that (using somthing the current directory) is actually what youre doing.

The empty string "" is the default parmeter of get_wml_location() though and most calls where its empty use it as that, we could split it into two functions one that takes a relative_dir parameter and one that doesn't, maybe that'd make it clearer

**Comment by soliton-** [comment]
> > Anyway, even without touching the weird get_wml_location logic the preprocessor can just call get_wml_location with the dir argument set to "." instead.
> 
> This wont work, if youpass `.` wesnoth will literally look for the file at "./included.cfg" (usually the working directory wesnoth was started from, or the folder the wesnoth.exe is i guess) which is most likely not the folder where where file.cfg is at.

It literally works that way as you can see if you hand in "." as the path.

> > Apparently this was done as a confusing bug fix: [1c927c8](https://github.com/wesnoth/wesnoth/commit/1c927c8407ec797ef9e9e58599a7b71e617fe864)
> 
> The "confusing bugfix" was a fix for a serious security problem that allowed umc code to read aribitiary files on the users computer.

If you could explain how it does that that’d be good to know. As far as I can tell the bug it’s trying to fix is not there. Even if it was it makes little sense to fix it by returning a nonsensical path. (The bug is getting fixed in 314425a also with returning a nonsensical path.)

> To make it clear: `get_wml_location ` is the central place that translates "wesnoth" paths to actual paths on the hard drive and must always return a path that's safe to read from.

There is some blacklisting (probably doesn’t belong here) based on mostly filename patterns and `current_dir` is checked for empty. I don’t see any checks to see if the return value points into the game or user data dir (which I assume is what you mean with safe paths).
